File: javav2/example_code/firehose/src/test/java/FirehoseTest.java
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import com.example.firehose.*;
import com.example.firehose.scenario.FirehoseScenario;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.Gson;
import org.junit.jupiter.api.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.auth.credentials.EnvironmentVariableCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.firehose.FirehoseClient;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
import java.io.*;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

/**
 * To run these integration tests, you must set the required values
 * in the config.properties file or AWS Secrets Manager.
 */
@TestInstance(TestInstance.Lifecycle.PER_METHOD)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public class FirehoseTest {
    private static final Logger logger = LoggerFactory.getLogger(FirehoseTest.class);
    private static FirehoseClient firehoseClient;
    private static String bucketARN = "";
    private static String roleARN = "";
    private static String newStream = "";
    private static String textValue = "";

    @BeforeAll
    public static void setUp() throws IOException {
        firehoseClient = FirehoseClient.builder()
                .region(Region.US_EAST_1)
                .build();

        // Get the values to run these tests from AWS Secrets Manager.
        Gson gson = new Gson();
        String json = getSecretValues();
        SecretValues values = gson.fromJson(json, SecretValues.class);
        bucketARN = values.getBucketARN();
        roleARN = values.getRoleARN();
        newStream = values.getNewStream() + java.util.UUID.randomUUID();
        textValue = values.getTextValue();

    }

    @Test
    @Tag("IntegrationTest")
    @Order(1)
    public void testCreateDeliveryStream() {
        assertDoesNotThrow(() -> {
            CreateDeliveryStream.createStream(firehoseClient, bucketARN, roleARN, newStream);
            CreateDeliveryStream.waitForStreamToBecomeActive(firehoseClient, newStream);
        });
        logger.info("Test 1 passed");
    }

    @Test
    @Tag("IntegrationTest")
    @Order(2)
    public void testPutRecord() throws IOException, InterruptedException {
        String jsonContent = FirehoseScenario.readJsonFile("sample_records.json");
        ObjectMapper objectMapper = new ObjectMapper();
        List<Map<String, Object>> sampleData = objectMapper.readValue(jsonContent, new TypeReference<>() {});

        // Process individual records.
        System.out.println("Processing individual records...");
        sampleData.subList(0, 100).forEach(record -> {
            try {
                FirehoseScenario.putRecord(record, newStream);
            } catch (Exception e) {
                System.err.println("Error processing record: " + e.getMessage());
            }
        });
        logger.info("Test 2 passed");
    }

   @Test
    @Tag("IntegrationTest")
    @Order(3)
    public void testListDeliveryStreams() {
        assertDoesNotThrow(() -> ListDeliveryStreams.listStreams(firehoseClient));
        logger.info("Test 3 passed");
    }

    @Test
    @Tag("IntegrationTest")
    @Order(4)
    public void testDeleteStream() {
        assertDoesNotThrow(() -> DeleteStream.delStream(firehoseClient, newStream));
        logger.info("Test 4 passed");
    }

    private static String getSecretValues() {
        SecretsManagerClient secretClient = SecretsManagerClient.builder()
                .region(Region.US_EAST_1)
                .build();
        String secretName = "test/firehose";

        GetSecretValueRequest valueRequest = GetSecretValueRequest.builder()
                .secretId(secretName)
                .build();

        GetSecretValueResponse valueResponse = secretClient.getSecretValue(valueRequest);
        return valueResponse.secretString();
    }

    @Nested
    @DisplayName("A class used to get test values from test/firehose (an AWS Secrets Manager secret)")
    class SecretValues {
        private String bucketARN;
        private String roleARN;
        private String newStream;

        private String textValue;

        public String getBucketARN() {
            return bucketARN;
        }

        public String getRoleARN() {
            return roleARN;
        }

        public String getNewStream() {
            return newStream;
        }

        public String getTextValue() {
            return textValue;
        }
    }
}

File: kotlin/services/firehose/src/test/kotlin/FirehoseTest.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import aws.sdk.kotlin.services.secretsmanager.SecretsManagerClient
import aws.sdk.kotlin.services.secretsmanager.model.GetSecretValueRequest
import com.google.gson.Gson
import com.kotlin.firehose.addStockTradeData
import com.kotlin.firehose.createStream
import com.kotlin.firehose.delStream
import com.kotlin.firehose.listStreams
import com.kotlin.firehose.putSingleRecord
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.MethodOrderer.OrderAnnotation
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestMethodOrder
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.UUID
import java.util.concurrent.TimeUnit

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(OrderAnnotation::class)
class FirehoseTest {
    private val logger: Logger = LoggerFactory.getLogger(FirehoseTest::class.java)
    private var bucketARN = ""
    private var roleARN = ""
    private var newStream = ""
    private var textValue = ""
    private var delStream = ""

    @BeforeAll
    fun setup() =
        runBlocking {
            // Get the values to run these tests from AWS Secrets Manager.
            val gson = Gson()
            val json: String = getSecretValues()
            val values = gson.fromJson(json, SecretValues::class.java)
            bucketARN = values.bucketARN.toString()
            roleARN = values.roleARN.toString()
            newStream = values.newStream.toString() + UUID.randomUUID()
            textValue = values.textValue.toString()
        }

    @Test
    @Order(1)
    fun createDeliveryStreamTest() =
        runBlocking {
            createStream(bucketARN, roleARN, newStream)
            logger.info("Test 1 passed")
        }

    @Test
    @Order(2)
    fun putRecordsTest() =
        runBlocking {
            // Wait for the resource to become available
            println("Wait 15 mins for resource to become available.")
            TimeUnit.MINUTES.sleep(15)
            putSingleRecord(textValue, newStream)
            logger.info("Test 2 passed")
        }

    @Test
    @Order(3)
    fun putBatchRecordsTest() =
        runBlocking {
            addStockTradeData(newStream)
            logger.info("Test 3 passed")
        }

    @Test
    @Order(4)
    fun listDeliveryStreamsTest() =
        runBlocking {
            listStreams()
            logger.info("Test 4 passed")
        }

    @Test
    @Order(5)
    fun deleteStreamTest() =
        runBlocking {
            delStream(newStream)
            logger.info("Test 5 passed")
        }

    private suspend fun getSecretValues(): String {
        val secretName = "test/firehose"
        val valueRequest =
            GetSecretValueRequest {
                secretId = secretName
            }
        SecretsManagerClient {
            region = "us-east-1"
        }.use { secretClient ->
            val valueResponse = secretClient.getSecretValue(valueRequest)
            return valueResponse.secretString.toString()
        }
    }

    @Nested
    @DisplayName("A class used to get test values from test/firehose (an AWS Secrets Manager secret)")
    internal class SecretValues {
        val bucketARN: String? = null
        val roleARN: String? = null
        val newStream: String? = null
        val textValue: String? = null
    }
}
